test(ivy): test ngcc with a CLI generated project

petebacondarwin opened the pull request:
This replaces https://github.com/angular/angular/pull/25869

petebacondarwin commented:
This is so close to working!
The problem now is that the `compileNgModuleFactory__POST_NGCC__` function is being shaken out of the flat file distribution files of the core library. I.E. `node_modules/@angular/core/fesm5/core.js` does not contain this function any more.

mary-poppins commented:
You can preview ec09b99 at https://pr26248-ec09b99.ngbuilds.io/.

mary-poppins commented:
You can preview 578e3cb at https://pr26248-578e3cb.ngbuilds.io/.

JoostK commented:
@petebacondarwin I created a PR two weeks ago for this 😁 #26071

petebacondarwin commented:
@JoostK you are a ⭐️

petebacondarwin commented:
Any other important PRs from you I should be aware of? :-)

petebacondarwin commented:
So after @JoostK 's fix, we are now breaking with `ngDevMode` is not defined...

JoostK commented:
> Any other important PRs from you I should be aware of? :-)

Nope, I don't think so. I was basically doing the same exercise as you, using a bare CLI 7 project together with ngcc. The FESM format fails because of the missing symbol, so I created #26071, then switched to using ESM and found the ngcc bugs you know of :-)

mary-poppins commented:
You can preview 3ddbbd3 at https://pr26248-3ddbbd3.ngbuilds.io/.

petebacondarwin committed:
fix(core): ensure that `ɵdefineNgModule` is available in flat-file formats

When compiling the flat-file version of the `@angular/core` we need to be aware
that we cannot rely upon imported names to access the ivy definition functions.
The compiler is already clever enough to use local function calls rather than
trying to add a namespaced import, but there is a problem if the local name of the
function is different to the exported name. This is the case for functions that
are not part of the public API, and so are exported under a barred-O private alias.

In `@angular/core` the only decorations in use are `@NgModule` and `@Injectable`.
There are no directives, components, pipes, etc.

Since `defineInjectable` is part of the public API of `@angular/core`, the compiler
is able to generate code that references the original non-barred-O version of the
function.

But the `defineNgModule` is not part of the public API and so the compiler must
generate code that refers to it by the private barred-O version of the function.

This commit imports and then re-exports this barred-O version of `defineModule` to
ensure that the symbol is available in the local scope of the flat-file versions of
the `@angular/core` library.

mary-poppins commented:
You can preview d50a8a4 at https://pr26248-d50a8a4.ngbuilds.io/.

petebacondarwin commented:
OK, so this is now blocked on a runtime error within Ivy.  We get `` error when rendering the `AppComponent`. Specifically when the `PercentPipe` tries to inject the `LOCALE_ID` into itself.

![screen shot 2018-10-05 at 14 34 52](https://user-images.githubusercontent.com/15655/46539626-ed8b6380-c8ae-11e8-9ddf-572710aa9b01.png)

When we get to `getOrCreateNodeInjector()` the `assertPreviousIsParent()` assertion fails.

![screen shot 2018-10-05 at 14 37 38](https://user-images.githubusercontent.com/15655/46539634-f3814480-c8ae-11e8-985a-b3924669c27c.png)

petebacondarwin committed:
test(ivy): ngcc - test compiling a CLI generated project

This integration test was created from a vanilla CLI generated
project with the following modifications:

* remove `PercentPipe` usage from `app.component.html`
  - these are not yet supported by ivy
* remove `ng test` from `package.json`
  - unit testing is not yet supported by ivy
* hard-code `ngDevMode` in `index.html`
  - the CLI does not yet set this correctly

mary-poppins commented:
You can preview 2a4f44b at https://pr26248-2a4f44b.ngbuilds.io/.

mary-poppins commented:
You can preview 8d4f710 at https://pr26248-8d4f710.ngbuilds.io/.

mary-poppins commented:
You can preview 7d4b823 at https://pr26248-7d4b823.ngbuilds.io/.

petebacondarwin committed:
refactor(ivy): ngcc - recombine flat and non-flat `Esm2015ReflectionHost`

Going forward we need to be able to do the same work on both
flat and non-flat module formats (such as computing arity and
transforming .d.ts files)

petebacondarwin committed:
fix(ivy): ngcc - support Angular Material package.json format

The Material project uses slightly different properties to the
core Angular project for specifying the different format entry-point.

This commit ensures that we map these properties correctly for both
types of project.

petebacondarwin committed:
refactor(ivy): ngcc - move typings rendering to `Renderer`

The rendering of typings is not specific to the package
format, so it doesn't make sense to put it in a specific
renderer.

As a result there is no real difference between esm5 and esm2015
renderers, so there is no point in having separate classes.

mary-poppins commented:
You can preview b508295 at https://pr26248-b508295.ngbuilds.io/.

petebacondarwin committed:
test(ivy): ngcc - test compiling the Angular Material library

* rename test helper script
* add material to the ngcc integration test
* add MatButton to ngcc integration test checks
* remove platform-server from ngcc integration test
  This package does not yet compile as it contains a package-private
  (internal) decorated class, which the ngcc compiler does not yet
  handle.

petebacondarwin committed:
feat(ivy): support separate .js and .d.ts trees when generating imports

The `NgModule` handler generates `R3References` for its declarations, imports,
exports, and bootstrap components, based on the relative import path
between the module and the classes it's referring to. This works fine for
compilation of a .ts Program inside ngtsc, but in ngcc the import needed
in the .d.ts file may be very different to the import needed between .js
files (for example, if the .js files are flattened and the .d.ts is not).

This commit introduces a new API in the `ReflectionHost` for extracting the
.d.ts version of a declaration, and makes use of it in the
`NgModuleDecorationHandler` to write a correct expression for the `NgModule`
definition type.

petebacondarwin committed:
fix(ivy): fix 'Module not found` error message

The message was dumping a serialized object instead of a
human readable name into the exception being thrown.

petebacondarwin committed:
fix(ivy): ngcc - skip missing formats rather than erroring

It is perfectly normal for some of the formats to be missing from
a package. We should not fail compilation for this.

petebacondarwin committed:
perf(ivy): ngcc - use flat file for dependency sorting if available

Previously we always used the non-flat format because we thought
that this was the one that would always be available.

It turns out that this is not the case and that only one of the flat and
non-flat formats may be available.

Therefore we should use whichever is available, defaulting to the flat
format if that exists, since that will be faster to parse.

mary-poppins commented:
You can preview bfe0295 at https://pr26248-bfe0295.ngbuilds.io/.

mary-poppins commented:
You can preview 30c7bf8 at https://pr26248-30c7bf8.ngbuilds.io/.

JoostK commented:
Could you drop 0aaf845 and 30c7bf8, they cancel each other out :)

petebacondarwin commented:
Good call.

mary-poppins commented:
You can preview cf9a01b at https://pr26248-cf9a01b.ngbuilds.io/.

petebacondarwin committed:
fix(ivy): ngcc - fixes to support compiling Material library

1) The `DecorationAnalyzer now analyzes all source files, rather than just
the entry-point files, which fixes #26183.
2) The `DecoratorAnalyzer` now runs all the `handler.analyze()`  calls
across the whole entry-point *before* running `handler.compile()`. This
ensures that dependencies between the decorated classes *within* an
entry-point are known to the handlers when running the compile process.
3) The `Renderer` now does the transformation of the typings (.d.ts) files
which allows us to support packages that only have flat format
entry-points better, and is faster, since we won't parse `.d.ts` files twice.

petebacondarwin committed:
perf(ivy): ngcc - only render .d.ts analysis when necessary

For each package entry-point there is only one format that
is used to compile the typings files (.d.ts). This will be
either esm2015 or fesm2015 (preferred). So we would not run
any dts processing in the renderer if we are not compiling
the appropriate format.

mary-poppins commented:
You can preview 54e773c at https://pr26248-54e773c.ngbuilds.io/.

mary-poppins commented:
You can preview a9e30ee at https://pr26248-a9e30ee.ngbuilds.io/.

mary-poppins commented:
You can preview 7737f54 at https://pr26248-7737f54.ngbuilds.io/.

mary-poppins commented:
You can preview c48a477 at https://pr26248-c48a477.ngbuilds.io/.

mary-poppins commented:
You can preview 53f4de5 at https://pr26248-53f4de5.ngbuilds.io/.

alxhub commented:
Code wise this lgtm, with a few comments.

Can you do me a favor and clean up the commits themselves? In particular:

* All the work to set up the CLI project can go into one commit at the end. Basically, the tests should be able to pass at every commit independently, so we should only add the test after all the ngcc work is done.
* Many of the commits are only single-line descriptions. Can you add a short paragraph for each one explaining the motivation behind the change? I typically write mine with a description of what the behavior was before, the problem it caused, and the action taken in the commit to fix it.

Other than that, this should be good to merge, nice work!

petebacondarwin commented on packages/core/src/core_render3_private_export.ts:
```diff
@@ -7,12 +7,13 @@
 
 // clang-format off
+import {defineNgModule} from './render3/index';
+export const ɵdefineNgModule = defineNgModule;
```
In `@angular/core` the only decorated classes are `NgModule` and `Injectable`. There are no directives, components, pipes, etc.

It turns out that `defineInjectable` is part of the public API of `@angular/core` so the compiler is able to generate code that references the original non-bared-o version of the function.

But the `defineNgModule` is not part of the public API and so the compiler must generate code that refers to it by the private bared-o version of the function.


I'll add this info to the commit message.

petebacondarwin commented on packages/core/src/core.ts:
```diff
@@ -11,6 +11,7 @@
  * @description
  * Entry point from which you should import all public core APIs.
  */
+export * from './core_render3_private_export';
```
It's an unnecessary change... sorry.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/packages/entry_point.ts:
```diff
@@ -41,11 +41,13 @@ interface EntryPointPackageJson {
   name: string;
   fesm2015?: string;
   fesm5?: string;
+  es2015?: string;  // if exists then it is actually FESM2015
```
It seems that this convention is pretty well followed, so unless we start finding situations where it is broken, I would prefer to avoid adding extra complexity (and processing effort) at this stage.
